### PR TITLE
LibPDF: Plug in the CCITT decoder

### PR DESCRIPTION
**Pull request by LucasChollet**
As mentioned somewhere on discord, there are issues in the CCITT decoder. So it doesn't make CCITT disappear from `--debugging-stats` but it's already enough to decode a lot of them.

Before:
```
Rendering of feature not supported: CCITTFaxDecode Filter Group 4 is unsupported, in 68 files, on 588 pages, 1826 times
[...]
767 files without issues (76.7%)
```

After:
```
Internal error while processing PDF file: CCITTDecoder: Corrupted stream, in 19 files, on 173 pages, 359 times
[...]
Internal error while processing PDF file: TIFFImageDecoderPlugin: CCITT codes encode for more than a line, in 10 files, on 49 pages, 63 times
[...]
804 files without issues (80.4%)
```

And with an image, before:
0000711_p2_before
After:
0000711_p2_after

@nico

### CODE REVIEW

File: Userland/Libraries/LibPDF/Filter.cpp [hunk 1]
```diff
@@ -275,7 +276,15 @@ PDFErrorOr<ByteBuffer> Filter::decode_run_length(ReadonlyBytes bytes)
     return TRY(Compress::PackBits::decode_all(bytes, OptionalNone {}, Compress::PackBits::CompatibilityMode::PDF));
 }
 
-PDFErrorOr<ByteBuffer> Filter::decode_ccitt(ReadonlyBytes, RefPtr<DictObject> decode_parms)
+static void handle_black_is_one(ByteBuffer& decoded, bool black_is_one)
```
Comment (nico) [inline Userland/Libraries/LibPDF/Filter.cpp:279]: very nit: i'd call this "invert_bits()" and move the `if (!black_is_1)` check to the calling site.

**Review by nico — APPROVED** [review]
Nice, LG! Some nits below, but as-is is fine too, up to you.

File: Userland/Libraries/LibPDF/Filter.cpp [hunk 2]
```diff
@@ -305,17 +306,15 @@ PDFErrorOr<ByteBuffer> Filter::decode_ccitt(ReadonlyBytes, RefPtr<DictObject> de
             damaged_rows_before_error = decode_parms->get_value(CommonNames::DamagedRowsBeforeError).get<int>();
     }
 
-    // FIXME: Do something with these.
-    (void)require_end_of_line;
-    (void)encoded_byte_align;
-    (void)columns;
-    (void)rows;
+    // FIXME: This parameter seems crucial when reading its description,
+    //        but I can't see any issue when ignoring it.
```
Comment (nico) [inline Userland/Libraries/LibPDF/Filter.cpp:318]: Given we still have

```
Internal error while processing PDF file: CCITTDecoder: Corrupted stream, in 9 files, on 30 pages, 31 times
```

I'm not sure "can't see" is correct. (Maybe it's not due to this, but we don't know.) Maybe "Figure out what to to do with it." instead?

**Comment by nico** [comment]
I went through a few PDFs to look at CCITTs, and they don't decode right (without diag) at least for these:

    Build/lagom/bin/pdf --render out.png /Users/thakis/Downloads/0000/0000031.pdf
    Build/lagom/bin/pdf --render out.png /Users/thakis/Downloads/0000/0000031.pdf --page 2
    # and all other pages of that PDF

    Build/lagom/bin/pdf --render out.png /Users/thakis/Downloads/0000/0000374.pdf
    # (and other pages)

    Build/lagom/bin/pdf --render out.png /Users/thakis/Downloads/0000/0000462.pdf --page 63
    # (also page 74, and possibly 81 95 121)

    Build/lagom/bin/pdf --render out.png /Users/thakis/Downloads/0000/0000650.pdf --page 19
    # (also page 22 30 31)
    # (the CCITT is used as the mask for another bitmap, which is how we get colors)

    Build/lagom/bin/pdf --render out.png /Users/thakis/Downloads/0000/0000425.pdf

Randomly trying documents, they seem to not show up much more often than they do. And not showing up with a diag is kind of worse than not showing up with a diag. Maybe it's easy to identify in which cases we can't decode yet and reject those with an error for now?

Also fine to merge as-is and figure out in tree if you prefer; up to you.

**Comment by nico** [comment]
…looks like the ones that don't decode right are exactly those with `row == 0` (which I think means "keep decoding rows while you have data"). So maybe we can keep rejecting those for now, until that's implemented?

**Comment by nico** [comment]
Adding `rows == 0` to the "Unsupported option for the CCITTFaxDecode Filter" line, this (and the other PR) bring us from 68 files with CCITTs that don't decode to 46 files with CCITTs that don't decode. A great progression, and as far as I can tell in line with reality :) So let's go with that, and then figure out `row == 0` later (ideally with a test case if possible, since that seems like something subtle).

**Comment by nico** [comment]
Here's some befores/afters, with #23305 applied as well.

Before:

Screenshot 2024-02-22 at 8 24 47 AM

After:

Screenshot 2024-02-22 at 8 00 11 AM

Before:

Screenshot 2024-02-22 at 8 24 51 AM

After:

Screenshot 2024-02-22 at 8 00 16 AM

Before:

Screenshot 2024-02-22 at 8 26 13 AM

After:

Screenshot 2024-02-22 at 8 26 55 AM

Before:

Screenshot 2024-02-22 at 8 25 26 AM

After:

Screenshot 2024-02-22 at 8 23 00 AM

Before:

Screenshot 2024-02-22 at 8 25 16 AM

After:

Screenshot 2024-02-22 at 8 01 50 AM

Before:

Screenshot 2024-02-22 at 8 25 49 AM

After:

Screenshot 2024-02-22 at 8 23 51 AM

(…and a whole bunch which are just completely empty before and a CCITT of a scanned page after.)

**Comment by nico** [comment]
(please just add `|| row == 0` to that reject if and implement real support in a follow-up. Incremental progress is good 🙂)

File: Userland/Libraries/LibPDF/Filter.cpp (hunk 2 above)
Comment (LucasChollet) [reply Userland/Libraries/LibPDF/Filter.cpp:318]: Oh I didn't see these :thinking:

File: Userland/Libraries/LibPDF/Filter.cpp [hunk 3]
```diff
@@ -305,20 +314,24 @@ PDFErrorOr<ByteBuffer> Filter::decode_ccitt(ReadonlyBytes, RefPtr<DictObject> de
...
-    (void)damaged_rows_before_error;
 
+    if (require_end_of_line || encoded_byte_align || damaged_rows_before_error > 0)
+        return Error::rendering_unsupported_error("Unsupported option for the CCITTFaxDecode Filter");
+
+    ByteBuffer decoded {};
     if (k < 0)
-        return Error::rendering_unsupported_error("CCITTFaxDecode Filter Group 4 is unsupported");
-    if (k == 0)
+        decoded = TRY(Gfx::CCITT::decode_ccitt_group4(bytes, columns, rows));
```
[the remark this answers is not in the document]
Comment (nico) [reply Userland/Libraries/LibPDF/Filter.cpp:324]: (Could also the github PR title, but meh, that doesn't go into the commit log.)

**Review by nico — APPROVED** [review]
lgtm \o/

File: Userland/Libraries/LibPDF/Filter.cpp [hunk 4]
```diff
@@ -305,20 +312,25 @@ PDFErrorOr<ByteBuffer> Filter::decode_ccitt(ReadonlyBytes, RefPtr<DictObject> de
             damaged_rows_before_error = decode_parms->get_value(CommonNames::DamagedRowsBeforeError).get<int>();
     }
 
-    // FIXME: Do something with these.
-    (void)require_end_of_line;
-    (void)encoded_byte_align;
-    (void)columns;
-    (void)rows;
+    // FIXME: This parameter seems crucial when reading its description, but we still
+    //        achieve to decode images with it. Figure out what to do with it.
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibPDF/Filter.cpp:316]: Nit: this comment is confusing me. Do you mean that we still manage to decode images without it?

Comment (nico) [reply Userland/Libraries/LibPDF/Filter.cpp:316]: I understood it to mean that we manage to decode (some?) images that have it set even though we ignore it.

**Review by ADKaster — APPROVED** [review]

**Pull request merged**